### DISCUSSION TITLE
Mocking MIDI devices for E2E testing

### DISCUSSION THREAD

**Discussion by jjeff** [djipco/webmidi#212] [Q&A]
I'm building an Electron app and using Playwright for E2E testing. I'm about to implement testing for the app's MIDI functionality (which uses WebMidi.js). I came across this old package which looks like it can mock MIDI devices: https://github.com/mohayonao/web-midi-test-api. However, I'm wondering if there's a more WebMidi.js-style of doing this? And/or does anyone have any suggestions for doing this with the WebMidi.js APIs?

I want to be able to mock a device and then be able to send MIDI messages from it... and potentially to it.

**Comment by djipco** [comment]
The way I'm doing it to unit test the library is to use the `midi` NPM package. It allows me to create virtual MIDI ports which I can then test the functionalities again. All my tests are performed in the Node.js environment. You can see examples in the `/test` directory.

At one point, I would love to have E2E testing in the browser with Playwright. Haven't had the time to look at this yet, though.
